README.md
# Updated

Managing state of asynchronous updates as a simple dart representation.

## Quickstart

A simple usage example of the `Update` state and `update` method.

```dart
import 'package:updated/updated.dart';

Future<int> requestValue() async {
  await Future.delayed(const Duration(seconds: 1));
  return 42;
}

Future<void> main() async {
  var value = Update<int>();

  // ---
  // Value : 32
  // Is Loading...
  // Details : Updating<int>(id: 28313824297, startedAt: 2020-11-23 16:57:04.298751, optimisticValue: 32)
  // ---
  // Value : 42
  // Is not loading : Updated<int>(id: 28313824297, updatedAt: 2020-11-23 16:57:05.315, previousUpdate: 42)
  // Details : Updated<int>(id: 28313824297, updatedAt: 2020-11-23 16:57:05.315, previousUpdate: 42)
  await for (var item in update(
    updater: requestValue,
    getUpdate: () => value,
    optimisticValue: 32,
  )) {
    print('---');
    value = item;
    item.mapValue(
      value: (value, isOptimistic) => print('Value : $value'),
      orElse: () => print('No value'),
    );
    item.mapLoading(
      loading: () => print('Is Loading...'),
      notLoading: () => print('Is not loading : $value'),
    );
    print('Details : $item');
  }

  // ---
  // Value : 32
  // Is Loading...
  // Details : Refreshing<int>(id: 28313797308, startedAt: 2020-11-23 16:56:38.327266, previousUpdate: Updated<int>(id: 28313797307, updatedAt: 2020-11-23 16:56:38.322668, previousUpdate: 42), optimisticValue: 32)
  // ---
  // Value : 42
  // Is not loading : Updated<int>(id: 28313797308, updatedAt: 2020-11-23 16:56:39.334887, previousUpdate: 42)
  // Details : Updated<int>(id: 28313797308, updatedAt: 2020-11-23 16:56:39.334887, previousUpdate: 42)
  await for (var item in update(
    updater: requestValue,
    getUpdate: () => value,
    optimisticValue: 32,
  )) {
    print('---');
    value = item;
    item.mapValue(
      value: (value, isOptimistic) => print('Value : $value'),
      orElse: () => print('No value'),
    );
    item.mapLoading(
      loading: () => print('Is Loading...'),
      notLoading: () => print('Is not loading : $value'),
    );
    print('Details : $item');
  }
}
```

## Usage

An update represents the lifecycle of a `T` value that can be loaded asynchronously.

![lifecycle](image.png)

Its initial state is `NotLoaded`. During its first update, it goes through the `Updating` state then `Updated` or `FailedUpdate` states, whether it is a success or not. During the following updates, it goes through the `Refreshing` state then `Updated` or `FailedRefresh` states, whether it is a success or not.

#### Initialization

To create an updated property, create an `Update<T>`.

```dart
final property = Update<int>();
```

#### Updating

The easiest way to update your `Update<T>` instance is by using the `update` function that returns a `Stream` with the sequence of updates.

```dart
var value = Update<int>();

final updateStream =  update(
    // The actual async operation
    updater: () async {
      // your async calls
    },
    // Make sure to always return the current value. This makes cancellation possible.
    getUpdate: () => value, 
    // This value will be available durint loading states
    optimisticValue: 32, 
    // To change the behaviour when an update is triggered but the previous isn't finished yet. 
    // If a previous update is cancelled then its stream is closed an no more events are emitted.
    override: UpdateOverride.cancelPrevious,
  ));
```

> If an `optimisticValue` is given, then its value is returned by the `Update.mapValue` method. This may be useful to display the result to the user beforehand if we're able to anticipate it (for example, by liking a tweet on twitter, we can fill the heart before receiving the result of the request that validates it).

> If a new update is triggered with the `cancelPrevious` option, then the previous update stream won't emit its final updates.

#### Mapping result

You can map the current state of an `Update<T>` instance with the `map`, `mapValue`, `mapLoading`, `mapError` methods.

```dart
@override
Widget build(BuildContext context)
 return state.products.mapValue(
      value: (value, isOptimistic) => ProductList(products: value),
      orElse: () => EmptyView(),
    );
}
```

```dart
@override
Widget build(BuildContext context)
 return state.products.mapError(
      error: (error, stackTrace) => ErrorView(error: error),
      orElse: () => ResulView(),
    );
}
```

```dart
@override
Widget build(BuildContext context)
 return state.products.mapLoading(
      loading: (error, stackTrace) => LoadingView(),
      notLoading: () => ResulView(),
    );
}
```


```dart
@override
Widget build(BuildContext context)
 return state.products.map(
      updated: (state) => UpdatedView(),
      failedRefresh: (state) => FailedView(),
      refreshing: (state) => LoadingView(),
      updating: (state) => LoadingView(),
      failedUpdate: (state) => FailedView(),
      notLoaded: (state) => NotLoadedView(),
    );
}
```

#### UpdateNotifier

An `UpdateNotifier<T>` is a mutable store for an `Update<T>` instance.

The [update] is updated by calling [execute].

```dart
final notifier = UpdateNotifier<int>();

final currentUpdate = notifier.update;

notifier.updateChanged.listen((update) {
    // ...
});

notifier.execute(
    // The actual async operation
    updater: () async {
      // your async calls
    },
    // This value will be available durint loading states
    optimisticValue: 32, 
    // To change the behaviour when an update is triggered but the previous isn't finished yet. 
    // If a previous update is cancelled then its stream is closed an no more events are emitted.
    override: UpdateOverride.cancelPrevious,
  ));
```

> If you're using patterns base on immutable data (like redux, MVU), using `update` function directly may be more appropriate.

CHANGELOG.md
## 0.1.10

- Added `mapUpdate` method.

## 0.1.9

- Added `Update.combine` function.

## 0.1.7

- Fixed `expirationDate`.

## 0.1.6

- Added `expirationDate` and `expirationDuration`.

## 0.1.5

- Fixed `maybeMap`.

## 0.1.4

- Added `maybeMap` and `hasSucceeded` to `Update`.

## 0.1.3

- Added default constructor for `Updated`.

## 0.1.2

- Added the `UpdateNotifier<T>`.

## 0.1.1

- Updated documentation.

## 0.1.0

- Initial version


test/future_test.dart
import 'package:meta/meta.dart';
import 'package:updated/updated.dart';
import 'package:updated/updated.dart' as u;
import 'package:test/test.dart';

class TestStore<T> {
  Update<T> value;
  TestStore.notUpdated() : value = Update<T>();

  TestStore.updating({
    @required int id,
  }) : value = Updating.fromNotLoaded(Update<T>(), id: id);

  TestStore.refreshing({
    @required int id,
    @required T value,
  }) : value = Refreshing.fromUpdated(
            Updated<T>.fromUpdating(
              Updating.fromNotLoaded(Update<T>(), id: id),
              value,
            ),
            id: id);

  TestStore.updated({
    @required int id,
    @required T value,
  }) : value = Updated<T>.fromUpdating(
          Updating.fromNotLoaded(Update<T>(), id: id),
          value,
        );

  Future<List<Update<T>>> update({
    @required Future<T> Function() updater,
    UpdateOverride override = UpdateOverride.ignore,
    T optimisticValue,
  }) async {
    final result = <Update<T>>[];
    await for (var item in u.update(
      updater: updater,
      getUpdate: () => value,
      override: override,
      optimisticValue: optimisticValue,
    )) {
      value = item;
      result.add(item);
    }
    return result;
  }
}

void main() {
  group('An update call', () {
    test('transitions from `NotLoaded` to `Updating` then `Updated`', () async {
      final store = TestStore.notUpdated();

      final updates = await store.update(
        updater: () async {
          await Future.delayed(const Duration(seconds: 1));
          return 42;
        },
      );

      expect(updates.length, 2);

      // Updating
      final updating = updates.first;
      expect(updating.isLoading, isTrue);
      expect(updating.hasValue, isFalse);
      expect(
          updating.map(
            updating: (_) => true,
            notLoaded: (_) => false,
            failedUpdate: (_) => false,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => false,
          ),
          isTrue);

      // Updated
      final updated = updates[1];
      expect(updated.isLoading, isFalse);
      expect(updated.hasValue, isTrue);
      expect(
          updated.map(
            updating: (_) => false,
            notLoaded: (_) => false,
            failedUpdate: (_) => false,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => true,
          ),
          isTrue);
      expect(updated.getValue(defaultValue: () => throw Exception), equals(42));
    });

    test('transitions from `NotLoaded` to `Updating` with an optimistic value',
        () async {
      final store = TestStore.notUpdated();

      final updates = await store.update(
        optimisticValue: 71,
        updater: () async {
          await Future.delayed(const Duration(seconds: 1));
          return 42;
        },
      );

      expect(updates.length, 2);

      // Updating
      final updating = updates.first;
      expect(updating.isLoading, isTrue);
      expect(updating.hasValue, isTrue);
      expect(
          updating.map(
            updating: (_) => true,
            notLoaded: (_) => false,
            failedUpdate: (_) => false,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => false,
          ),
          isTrue);
      expect(
          updating.getValue(defaultValue: () => throw Exception), equals(71));

      // Updated
      final updated = updates[1];
      expect(updated.isLoading, isFalse);
      expect(updated.hasValue, isTrue);
      expect(
          updated.map(
            updating: (_) => false,
            notLoaded: (_) => false,
            failedUpdate: (_) => false,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => true,
          ),
          isTrue);
      expect(updated.getValue(defaultValue: () => throw Exception), equals(42));
    });

    test('transitions from `NotLoaded` to `Updating` then `FailedUpdate`',
        () async {
      final store = TestStore.notUpdated();

      final updates = await store.update(
        updater: () async {
          await Future.delayed(const Duration(seconds: 1));
          throw Exception('TEST');
        },
      );

      expect(updates.length, 2);

      // Updating
      final updating = updates.first;
      expect(updating.isLoading, isTrue);
      expect(updating.hasValue, isFalse);
      expect(
          updating.map(
            updating: (_) => true,
            notLoaded: (_) => false,
            failedUpdate: (_) => false,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => false,
          ),
          isTrue);

      // FailedUpdate
      final failed = updates[1];
      expect(failed.hasFailed, isTrue);
      expect(failed.isLoading, isFalse);
      expect(failed.hasValue, isFalse);
      expect(
          failed.map(
            updating: (_) => false,
            notLoaded: (_) => false,
            failedUpdate: (_) => true,
            refreshing: (_) => false,
            failedRefresh: (_) => false,
            updated: (_) => false,
          ),
          isTrue);
      expect(
          failed.mapError(
            error: (error, stackTrace) => error.message,
            orElse: () => throw Exception,
          ),
          equals('TEST'));
    });
  });

  test('transitions from `Updated` to `Refreshing` then `Updated`', () async {
    final store = TestStore.updated(id: 1, value: 42);

    // Starting a refresh
    final updates = await store.update(
      updater: () async {
        await Future.delayed(const Duration(seconds: 1));
        return 24;
      },
    );

    expect(updates.length, 2);

    // Refreshing
    final refreshing = updates.first;
    expect(refreshing.isLoading, isTrue);
    expect(refreshing.hasValue, isTrue);
    expect(
        refreshing.map(
          updating: (_) => false,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => true,
          failedRefresh: (_) => false,
          updated: (_) => false,
        ),
        isTrue);
    expect(
        refreshing.getValue(defaultValue: () => throw Exception), equals(42));

    // Updated
    final refreshed = updates[1];
    expect(refreshed.isLoading, isFalse);
    expect(refreshed.hasValue, isTrue);
    expect(
        refreshed.map(
          updating: (_) => false,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => false,
          failedRefresh: (_) => false,
          updated: (_) => true,
        ),
        isTrue);
    expect(refreshed.getValue(defaultValue: () => throw Exception), equals(24));
  });

  test('does nothing from `Updating` with ignore override', () async {
    final store = TestStore.updating(id: 1);

    // Starting a refresh
    final updates = await store.update(
      override: UpdateOverride.ignore,
      updater: () async {
        await Future.delayed(const Duration(seconds: 1));
        return 24;
      },
    );

    expect(updates.length, 0);

    // Updating
    final updating = store.value;
    expect(updating.isLoading, isTrue);
    expect(updating.hasValue, isFalse);
    expect(
        updating.map(
          updating: (_) => true,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => false,
          failedRefresh: (_) => false,
          updated: (_) => false,
        ),
        isTrue);
  });

  test('does nothing from `Refreshing` with ignore override', () async {
    final store = TestStore.refreshing(id: 1, value: 42);

    // Starting a refresh
    final updates = await store.update(
      override: UpdateOverride.ignore,
      updater: () async {
        await Future.delayed(const Duration(seconds: 1));
        return 24;
      },
    );

    expect(updates.length, 0);

    // Refreshing
    final refreshing = store.value;
    expect(refreshing.isLoading, isTrue);
    expect(refreshing.hasValue, isTrue);
    expect(
        refreshing.map(
          updating: (_) => false,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => true,
          failedRefresh: (_) => false,
          updated: (_) => false,
        ),
        isTrue);
    expect(
        refreshing.getValue(defaultValue: () => throw Exception), equals(42));
  });

  test('cancels `Updating` when cancelPrevious override', () async {
    final store = TestStore.notUpdated();

    final cancelledUpdatesFuture = store.update(
      updater: () async {
        await Future.delayed(const Duration(seconds: 3));
        return 24;
      },
    );

    await Future.delayed(const Duration(seconds: 1));

    final overridingUpdatesFuture = store.update(
      override: UpdateOverride.cancelPrevious,
      updater: () async {
        await Future.delayed(const Duration(seconds: 1));
        return 42;
      },
    );

    final cancelledUpdates = await cancelledUpdatesFuture;
    final overridingUpdates = await overridingUpdatesFuture;

    expect(cancelledUpdates.length, 1);

    // Cancelled : Updating
    final cancelUpdating = cancelledUpdates.first;
    expect(cancelUpdating.isLoading, isTrue);
    expect(cancelUpdating.hasValue, isFalse);
    expect(
        cancelUpdating.map(
          updating: (_) => true,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => false,
          failedRefresh: (_) => false,
          updated: (_) => false,
        ),
        isTrue);

    expect(overridingUpdates.length, 2);

    // Updating
    final updating = overridingUpdates.first;
    expect(updating.isLoading, isTrue);
    expect(updating.hasValue, isFalse);
    expect(
        updating.map(
          updating: (_) => true,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => false,
          failedRefresh: (_) => false,
          updated: (_) => false,
        ),
        isTrue);

    // Updated
    final updated = overridingUpdates[1];
    expect(updated.isLoading, isFalse);
    expect(updated.hasValue, isTrue);
    expect(
        updated.map(
          updating: (_) => false,
          notLoaded: (_) => false,
          failedUpdate: (_) => false,
          refreshing: (_) => false,
          failedRefresh: (_) => false,
          updated: (_) => true,
        ),
        isTrue);
    expect(updated.getValue(defaultValue: () => throw Exception), equals(42));
  });
}


lib/src/future.dart
import 'package:meta/meta.dart';

import 'update.dart';

/// Describes the behaviour of an update when one is already running or expired.
abstract class UpdateOverride {
  const UpdateOverride._();

  /// If a new update is requested and there already an ongoing update, the new update is ignored
  /// and the previous one continues.
  static const UpdateOverride ignore = _UpdateOverrideIgnore();

  /// If a new update is requested and there already an ongoing update, the previous one is cancelled
  /// and the new update starts.
  static const UpdateOverride cancelPrevious = _UpdateOverrideCancelPrevious();
}

/// Starts a sequence of updates by running the [updater] and raising a new [Update] after each step of its execution.
///
/// At each step of the updater execution, the [getUpdate] will be used to check if the task has been cancelled.
/// If so, the resulting stream will end without any new update.
///
/// If an update has already been started from the item returned by [getUpdate], then the behaviour is controlled
/// by the [override] parameters. The new update can whether be ignore, or cancel the previous execution.
///
/// An [optimisticValue] can be given to display an anticipated result during the loading phase.
///
/// If [expirationDate] is given, and the update has already been updated, then a new update is triggered only if the
/// the given date is before [DateTime.now].
///
/// If [expirationDuration] is given, and no [expirationDate] is given, then [expirationDate] becomes the last updated
/// date with the duration added to it.
Stream<Update<T>> update<T>({
  @required Future<T> Function() updater,
  @required Update<T> Function() getUpdate,
  UpdateOverride override = UpdateOverride.ignore,
  T optimisticValue,
  DateTime expirationDate,
  Duration expirationDuration,
}) {
  assert(getUpdate != null);
  assert(updater != null);
  return getUpdate().map(
    notLoaded: (state) async* {
      final updating = Updating.fromNotLoaded(
        state,
        id: _createId(),
        optimisticValue: optimisticValue,
      );
      yield updating;
      try {
        final result = await updater();
        if (!getUpdate().isCancelled(updating.id)) {
          yield Updated.fromUpdating(updating, result);
        }
      } catch (error, stackTrace) {
        if (!getUpdate().isCancelled(updating.id)) {
          yield FailedUpdate.fromUpdating(
            updating,
            error: error,
            stackTrace: stackTrace,
          );
        }
      }
    },
    updating: (state) async* {
      if (override is _UpdateOverrideCancelPrevious) {
        final updating = Updating.cancelling(
          state,
          id: _createId(),
          optimisticValue: optimisticValue,
        );
        yield updating;
        try {
          final result = await updater();
          if (!getUpdate().isCancelled(updating.id)) {
            yield Updated.fromUpdating(updating, result);
          }
        } catch (error, stackTrace) {
          if (!getUpdate().isCancelled(updating.id)) {
            yield FailedUpdate.fromUpdating(
              updating,
              error: error,
              stackTrace: stackTrace,
            );
          }
        }
      }
    },
    updated: (state) async* {
      if (expirationDuration != null) {
        expirationDate ??= state.updatedAt.add(expirationDuration);
      }
      final isExpired =
          expirationDate == null || expirationDate.isBefore(DateTime.now());
      if (isExpired) {
        final updating = Refreshing.fromUpdated(
          state,
          id: _createId(),
          optimisticValue: optimisticValue,
        );
        yield updating;
        try {
          final result = await updater();
          if (!getUpdate().isCancelled(updating.id)) {
            yield Updated.fromRefreshing(updating, result);
          }
        } catch (error, stackTrace) {
          if (!getUpdate().isCancelled(updating.id)) {
            yield FailedRefresh.fromRefreshing(
              updating,
              error: error,
              stackTrace: stackTrace,
            );
          }
        }
      }
    },
    failedRefresh: (state) async* {
      final updating = Refreshing.fromFailed(
        state,
        id: _createId(),
        optimisticValue: optimisticValue,
      );
      yield updating;
      try {
        final result = await updater();
        if (!getUpdate().isCancelled(updating.id)) {
          yield Updated.fromRefreshing(updating, result);
        }
      } catch (error, stackTrace) {
        if (!getUpdate().isCancelled(updating.id)) {
          yield FailedRefresh.fromRefreshing(
            updating,
            error: error,
            stackTrace: stackTrace,
          );
        }
      }
    },
    failedUpdate: (state) async* {
      final updating = Updating.fromFailed(
        state,
        id: _createId(),
        optimisticValue: optimisticValue,
      );
      yield updating;
      try {
        final result = await updater();
        if (!getUpdate().isCancelled(updating.id)) {
          yield Updated.fromUpdating(updating, result);
        }
      } catch (error, stackTrace) {
        if (!getUpdate().isCancelled(updating.id)) {
          yield FailedUpdate.fromUpdating(
            updating,
            error: error,
            stackTrace: stackTrace,
          );
        }
      }
    },
    refreshing: (state) async* {
      if (override is _UpdateOverrideCancelPrevious) {
        final updating = Refreshing.cancelling(
          state,
          id: _createId(),
          optimisticValue: optimisticValue,
        );
        yield updating;
        try {
          final result = await updater();
          if (!getUpdate().isCancelled(updating.id)) {
            yield Updated.fromRefreshing(updating, result);
          }
        } catch (error, stackTrace) {
          if (!getUpdate().isCancelled(updating.id)) {
            yield FailedRefresh.fromRefreshing(
              updating,
              error: error,
              stackTrace: stackTrace,
            );
          }
        }
      }
    },
  );
}

extension _UpdateExtensions<T> on Update<T> {
  /// Test if the [id] is the same as the update one.
  bool isCancelled(int id) {
    return map(
      failedRefresh: (state) => state.id != id,
      failedUpdate: (state) => state.id != id,
      updated: (state) => state.id != id,
      refreshing: (state) => state.id != id,
      updating: (state) => state.id != id,
      notLoaded: (state) => false,
    );
  }
}

/// Create a new unique identifiers which will be associated to a new update.
int _createId() => _lastId++;

/// The unique identifiers associated to updates.
int _lastId = DateTime.now().millisecondsSinceEpoch -
    DateTime(2020, 1, 1).millisecondsSinceEpoch;

class _UpdateOverrideIgnore extends UpdateOverride {
  const _UpdateOverrideIgnore() : super._();
}

class _UpdateOverrideCancelPrevious extends UpdateOverride {
  const _UpdateOverrideCancelPrevious() : super._();
}


lib/src/notifier.dart
import 'dart:async';

import 'package:meta/meta.dart';

import 'future.dart' as future;
import 'update.dart';

/// A notifier is a store for a given [Update<T>].
///
/// The [update] is updated by calling [execute].
///
/// If you're using patterns base on immutable data (like redux, MVU), using `update` function directly
/// may be more appropriate.
class UpdateNotifier<T> {
  /// Create a new notifier.
  ///
  /// If an [initialValue] is given, the [update] is initialized as an [Updated<T>], else
  /// the value is initialized as a [NotUpdated<T>].
  UpdateNotifier({
    T initialValue,
  }) : _update = initialValue == null
            ? Update<T>()
            : Updated<T>.fromUpdating(
                Updating.fromNotLoaded(
                  NotLoaded<T>(),
                  id: 0,
                ),
                initialValue,
              );

  /// Indicates a new change of [update].
  ///
  /// Changes are emmited by calling the [execute] method.
  Stream<Update<T>> get updateChanged => _updateChanged.stream;

  /// Get the current value.
  Update<T> get update => _update;

  final StreamController<Update<T>> _updateChanged =
      StreamController<Update<T>>.broadcast();

  Update<T> _update;

  /// Starts a sequence of updates by running the [updater] and raising a new [Update] after each step of its execution.
  ///
  /// If an update has already been started from the current [update], then the behaviour is controlled
  /// by the [override] parameters. The new update can whether be ignore, or cancel the previous execution.
  ///
  /// An [optimisticValue] can be given to display an anticipated result during the loading phase.
  void execute({
    @required Future<T> Function() updater,
    future.UpdateOverride override = future.UpdateOverride.ignore,
    T optimisticValue,
  }) async {
    await for (var item in future.update(
      getUpdate: () => _update,
      updater: updater,
      override: override,
      optimisticValue: optimisticValue,
    )) {
      _update = item;
      _updateChanged.add(item);
    }
  }

  /// Release all resource (like stream listeners).
  void dispose() {
    _updateChanged.close();
  }
}


example/example.dart
import 'package:updated/updated.dart';

Future<int> requestValue() async {
  await Future.delayed(const Duration(seconds: 1));
  return 42;
}

Future<void> main() async {
  var value = Update<int>();

  // ---
  // Value : 32
  // Is Loading...
  // Details : Updating<int>(id: 28313824297, startedAt: 2020-11-23 16:57:04.298751, optimisticValue: 32)
  // ---
  // Value : 42
  // Is not loading : Updated<int>(id: 28313824297, updatedAt: 2020-11-23 16:57:05.315, previousUpdate: 42)
  // Details : Updated<int>(id: 28313824297, updatedAt: 2020-11-23 16:57:05.315, previousUpdate: 42)
  await for (var item in update(
    updater: requestValue,
    getUpdate: () => value,
    optimisticValue: 32,
  )) {
    print('---');
    value = item;
    item.mapValue(
      value: (value, isOptimistic) => print('Value : $value'),
      orElse: () => print('No value'),
    );
    item.mapLoading(
      loading: () => print('Is Loading...'),
      notLoading: () => print('Is not loading : $value'),
    );
    print('Details : $item');
  }

  // ---
  // Value : 32
  // Is Loading...
  // Details : Refreshing<int>(id: 28313797308, startedAt: 2020-11-23 16:56:38.327266, previousUpdate: Updated<int>(id: 28313797307, updatedAt: 2020-11-23 16:56:38.322668, previousUpdate: 42), optimisticValue: 32)
  // ---
  // Value : 42
  // Is not loading : Updated<int>(id: 28313797308, updatedAt: 2020-11-23 16:56:39.334887, previousUpdate: 42)
  // Details : Updated<int>(id: 28313797308, updatedAt: 2020-11-23 16:56:39.334887, previousUpdate: 42)
  await for (var item in update(
    updater: requestValue,
    getUpdate: () => value,
    optimisticValue: 32,
  )) {
    print('---');
    value = item;
    item.mapValue(
      value: (value, isOptimistic) => print('Value : $value'),
      orElse: () => print('No value'),
    );
    item.mapLoading(
      loading: () => print('Is Loading...'),
      notLoading: () => print('Is not loading : $value'),
    );
    print('Details : $item');
  }
}
